--- heimdall-1.3.1/heimdall/Makefile.am.bero	2012-09-11 19:17:01.307997342 +0200
+++ heimdall-1.3.1/heimdall/Makefile.am	2012-09-11 19:17:50.032249623 +0200
@@ -21,14 +21,6 @@ heimdall_LDADD = $(DEPS_LIBS) $(STATIC_L
 if LINUXTARGET
 udevrulesdir = /lib/udev/rules.d
 udevrules_DATA = 60-heimdall-galaxy-s.rules
-
-install-data-hook:
-if UDEVADM
-	sudo udevadm control --reload_rules
-else
-	sudo service udev restart
-endif
-
 endif
 
 dist_noinst_SCRIPTS = autogen.sh
